### ISSUE TITLE
Support passing paths to `--watch` flag of `deno test`

### ISSUE THREAD

**Issue by lionel-rowe** [denoland/deno#21704]
Currently, the following works as expected:

```sh
deno run --watch=watched-file.ts other-file.ts
```

However, the following fails:

```sh
deno test --watch=watched-file.ts other-file.ts
# error: unexpected value 'watched-file.ts' for '--watch' found; no more were expected
```

It seems like there isn't currently any way to customize the files that watched with the `test` command. It's also somewhat unexpected that `--watch` wouldn't work identically between `run` vs `test` (but I guess it might need to work somewhat differently due to multiple entrypoints?)

Motivating use case: test suite includes some doc tests that run code blocks found in markdown files, so the tests should be re-run on changes to the markdown files as well as ts/js files.

**Comment by bartlomieju** [comment]
This seems very reasonable and we should support it. It should be straightforward to implement following the same pattern as for `deno run` so marking as good first issue.

**Comment by RyanClementsHax** [comment]
I'm considering helping out here. Is this what you mean by straightforward or is there something I'm missing.

- [Argument parsing](https://github.com/denoland/deno/blob/261f32ef651a6515fbac664302adcfbe34a04372/cli/args/flags.rs#L2206) needs to be updated
- The [TestFlags struct](https://github.com/denoland/deno/blob/261f32ef651a6515fbac664302adcfbe34a04372/cli/args/flags.rs#L278) should be updated to use the same type as the [RunFlags struct](https://github.com/denoland/deno/blob/261f32ef651a6515fbac664302adcfbe34a04372/cli/args/flags.rs#L221)
- The [watch_paths()](https://github.com/denoland/deno/blob/261f32ef651a6515fbac664302adcfbe34a04372/cli/args/mod.rs#L1434) function needs to be updated to handle this case since it is used in the [test worker for watched runs](https://github.com/denoland/deno/blob/261f32ef651a6515fbac664302adcfbe34a04372/cli/tools/test/mod.rs#L1266)

**Comment by bartlomieju** [comment]
@RyanClementsHax that's correct. Feel free to open a PR and ping me there if you get stuck anywhere.

**Comment by RyanClementsHax** [comment]
Haven't forgotten about this!

**Comment by RyanClementsHax** [comment]
I'm still implementing this, but I'd figure I'd ask about some friction I was seeing as it took up all my "open source time" last week to solve.

Last week, I was coming by some issues building off of main and [1.39.2](https://github.com/denoland/deno/releases/tag/v1.39.2). Running `cargo clean && cargo build -vv` failed with being unable to compile `deno_runtime`. It didn't give me too much error info to work off of, but are tools/techniques to pinpoint problems when building `deno_runtime`?

The problem seems to be solved now, but I'm just looking to see what other people turn to when debugging rust build problems. Normally the compiler errors are helpful, but this time I didn't find them very helpful.

**Comment by ericlery** [comment]
> The problem seems to be solved now, but I'm just looking to see what other general things people turn to when debugging rust build problems for deno.

I've run into some issues myself first time building Deno. In [this issue](https://github.com/denoland/deno/issues/15357#issuecomment-1866503275), I've talk about the way I managed to build without errors, and how I got informations to debug what's happening during the build.

**Comment by RyanClementsHax** [comment]
Thanks for that. It's always nice to see how others are developing in deno

**Comment by RyanClementsHax** [comment]
Ok I have this mostly developed. I'm just not sure what to do about [this line](https://github.com/denoland/deno/blob/main/cli/tools/test/mod.rs#L1301) that includes the entire project directory when the `--watch` flag is passed.

@bartlomieju should the behavior be the following?
1. If `--watch` is passed, the root directory is watched as before
2. If one or more `--watch=some_file.ts` is passed, the root directory should not be watched, only the files explicitly mentioned, the entry point, and all files transitively or directly imported by the entry point

**Comment by RyanClementsHax** [comment]
I know you're busy. Just wanted to let you know that I'm still ready to continue work whenever you can get back to the questions above.

**Comment by muddlebee** [comment]
@lionel-rowe @bartlomieju we can close this. its working as expected

this is the correct format
```
deno test --watch main_test.ts 
```
rather than
```
deno test --watch=main_test.ts 

```

![image](https://github.com/denoland/deno/assets/8139783/cd3204b6-d1b6-44d2-970c-eb9ff9e73c14)

**Comment by lionel-rowe** [comment]
> @lionel-rowe @bartlomieju we can close this. its working as expected
> ```sh
> deno test --watch main_test.ts 
> ```

@muddlebee that watches `main_test.ts` and any of its dependencies but it doesn't watch any other files.

> Motivating use case: test suite includes some doc tests that run code blocks found in markdown files, so the tests should be re-run on changes to the markdown files as well as ts/js files.

Here's a repro
https://github.com/lionel-rowe/deno-test-watch-repro/blob/main/deno.jsonc

**Comment by muddlebee** [comment]
thank you. I understand. I will look into this.

**Comment by zeluspudding** [comment]
Has anyone figure out how to rerun tests when some non-test files are updated? Have been trying to get the watcher to rerun my tests when my main files are updated but there's seemingly no combination of flags I can pass in that makes it work. Am wanting exactly what @lionel-rowe outlined:

```
deno test -A --watch=docs.md                       *.test.ts
#                    ^ files to watch but not run  ^ files to run (and watch)
```

**Comment by chengyixu** [comment]
Hi! I'd like to work on this.

**Approach:** The `deno test --watch` flag should accept file/directory paths like `deno run --watch` does. The fix is in the CLI argument parsing for the `test` subcommand — update the `--watch` flag definition to accept an optional value (path list), mirroring the `run` subcommand's implementation. The watcher would then monitor both the test files AND the extra paths provided. I'd also add tests verifying that changes to specified paths trigger re-runs.
**Timeline:** 2-3 days
**Background:** Node.js/TypeScript developer, 35+ npm packages, familiar with CLI arg parsing patterns.

— @chengyixu
